crates/extensions/c8y_mapper_ext/src/json.rs

```diff
@@ -49,12 +51,79 @@ fn from_thin_edge_json_with_timestamp(
     timestamp: OffsetDateTime,
     entity: &CloudEntityMetadata,
     m_type: &str,
+    units: Option<&Units>,
 ) -> Result<String, CumulocityJsonError> {
-    let mut serializer = serializer::C8yJsonSerializer::new(timestamp, entity, m_type);
+    let mut serializer = serializer::C8yJsonSerializer::new(timestamp, entity, m_type, units);
     parse_str(input, &mut serializer)?;
     Ok(serializer.into_string()?)
 }
 
+/// Units used for measurements of a given type
+#[derive(Default)]
+pub struct Units {
+    units: HashMap<String, String>,
+    groups: HashMap<String, Units>,
+}
+
+impl Units {
+    /// An empty set of measurement units
+    ///
+    /// This is the default when no measurement metadata is published for a measurement topic
+    pub fn new() -> Units {
+        Units {
+            units: HashMap::new(),
+            groups: HashMap::new(),
+        }
+    }
+
+    /// True if not units are actually defined
+    pub fn is_empty(&self) -> bool {
+        self.units.is_empty() && self.groups.is_empty()
+    }
+
+    /// Measurement units as defined by metadata published on a measurement topic
+    pub fn from_metadata(meta: serde_json::Value) -> Self {
+        let mut units = Units::new();
+        if let serde_json::Value::Object(map) = meta {
+            for (k, v) in map {
+                units.set_unit(k, v);
+            }
+        }
+        units
+    }
+
+    pub fn set_unit(&mut self, measurement: String, meta: serde_json::Value) {
+        if let serde_json::Value::String(unit) = meta {
+            // "Temperature": "°C"
+            self.units.insert(measurement, unit);
```

Comment:
Not quite sure about this case where we assume the value to be units, if we expect `meta` to be expanded beyond units in future. But, if we're pretty confident that this is the most reasonable "default" `meta` for a measurement, then this is fine.